slackbot.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2018 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"fmt"
+	"time"
+
+	log "github.com/gravitational/logrus"
+	"github.com/gravitational/trace"
+
+	pagerduty "github.com/PagerDuty/go-pagerduty"
+	"github.com/shomali11/slacker"
+)
+
+// Init is called upon Bot creation (first startup)
+func Init(config *Config) {
+	log.Println("Connected!")
+	client := pagerduty.NewClient(config.PagerDuty.APIKey)
+	var opts pagerduty.GetScheduleOptions
+	if schedule, err := client.GetSchedule(config.PagerDuty.Schedule, opts); err != nil {
+		trace.Wrap(err)
+	} else {
+		log.Println("Configured schedule is \"" + schedule.Name + "\" with ID: " + config.PagerDuty.Schedule)
+	}
+}
+
+// Err function is used to handle all Error reported by the Bot
+func Err(err string) {
+	fmt.Println(err)
+}
+
+// Emergency is used to open Emergency Incidents on PagerDuty
+func Emergency(request slacker.Request, response slacker.ResponseWriter, config *Config) {
+	client := pagerduty.NewClient(config.PagerDuty.APIKey)
+	var scheduleOpts pagerduty.GetScheduleOptions
+	if schedule, err := client.GetSchedule(config.PagerDuty.Schedule, scheduleOpts); err != nil {
+		trace.Wrap(err)
+	} else {
+		log.Println("Opening incident on schedule \"" + schedule.Name +
+			"\"/" + config.PagerDuty.Schedule)
+	}
+
+	newIncident := pagerduty.CreateIncidentOptions{
+		Type: "incident",
+		Title: "Incident opened by " + config.CustomerName +
+			" via Slack/@" + config.Slack.BotUsername,
+	}
+
+	newIncidentBody := pagerduty.APIDetails{
+		Type:    "incident_body",
+		Details: request.Param("msg"),
+	}
+	newIncident.Body = newIncidentBody
```

Comment:
Do not allocate memory for new variable, use `newIncident.Body = pagerduty.APIDetails{...}` instead.

Reply:
Fixed.